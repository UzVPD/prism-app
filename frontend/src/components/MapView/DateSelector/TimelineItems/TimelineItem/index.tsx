import { WithStyles, createStyles, withStyles } from '@material-ui/core';
import React, { memo, useCallback } from 'react';
import 'react-datepicker/dist/react-datepicker.css';
import { DateItem, DateRangeType } from '../../../../../config/types';
import { datesAreEqualWithoutTime } from '../../../../../utils/date-utils';

const TimelineItem = memo(
  ({
    classes,
    concatenatedLayers,
    currentDate,
    index,
    clickDate,
    dateItemStyling,
  }: TimelineItemProps) => {
    const handleClick = useCallback(
      (dateIndex: number) => {
        return () => {
          clickDate(dateIndex);
        };
      },
      [clickDate],
    );

    return (
      <>
        {concatenatedLayers.map(
          (layerDates: DateItem[], layerIndex: number) => {
            // TODO: fix not really efficient algorithm
            const matchingDateItemInLayer = layerDates.find(f =>
              datesAreEqualWithoutTime(f.displayDate, currentDate.value),
            );

            if (!matchingDateItemInLayer) {
              return null;
            }
            return (
              <React.Fragment key={Math.random()}>
                {layerIndex !== 0 && matchingDateItemInLayer.isStartDate && (
                  <div
                    className={`${dateItemStyling[layerIndex].layerDirectionClass} ${classes.layerDirectionBase}`}
                  />
                )}
                {layerIndex !== 0 && matchingDateItemInLayer.isEndDate && (
                  <div
                    className={`${dateItemStyling[layerIndex].layerDirectionClass} ${classes.layerDirectionBase} ${classes.layerDirectionBackwardBase}`}
                  />
                )}

                <div
                  className={dateItemStyling[layerIndex].class}
                  role="presentation"
                  onClick={handleClick(index)}
                />
              </React.Fragment>
            );
          },
        )}
      </>
    );
  },
);

const styles = () =>
  createStyles({
    layerDirectionBase: {
      display: 'block',
      position: 'absolute',
      borderTop: '8px solid transparent',
      borderBottom: '8px solid transparent',
      height: '0px',
      zIndex: 1,
    },

    layerDirectionBackwardBase: {
      transform: 'rotate(180deg)',
    },
  });

export interface TimelineItemProps extends WithStyles<typeof styles> {
  concatenatedLayers: DateItem[][];
  currentDate: DateRangeType;
  index: number;
  clickDate: (arg: number) => void;
  dateItemStyling: {
    class: string;
    color: string;
    layerDirectionClass?: string;
  }[];
}

export default withStyles(styles)(TimelineItem);
