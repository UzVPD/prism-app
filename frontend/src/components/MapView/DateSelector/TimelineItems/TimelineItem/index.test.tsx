import React from 'react';
import { render } from '@testing-library/react';
import TimelineItem, { TimelineItemProps } from '.';

test('TimelineItem renders as expected', () => {
  // Arrange
  const currentDateTime = new Date().getTime();
  const props: Omit<TimelineItemProps, 'classes'> = {
    clickDate: () => {},
    concatenatedLayers: [
      [{ displayDate: currentDateTime, queryDate: currentDateTime }],
      [
        {
          displayDate: currentDateTime,
          queryDate: currentDateTime,
        },
        {
          displayDate: currentDateTime,
          queryDate: currentDateTime,
          isStartDate: true,
        },
      ],
    ],
    currentDate: {
      value: currentDateTime,
      label: '',
      month: '',
      isFirstDay: false,
      date: '',
    },
    index: 1,
    dateItemStyling: [
      { class: 'intersectionDate', color: 'White' },
      {
        class: 'layerOneDate',
        color: '',
        layerDirectionClass: 'layerDirectionClass',
      },
    ],
  };

  // Act
  const { container } = render(<TimelineItem {...props} />);

  // Assert
  expect(container).toMatchSnapshot();
});
