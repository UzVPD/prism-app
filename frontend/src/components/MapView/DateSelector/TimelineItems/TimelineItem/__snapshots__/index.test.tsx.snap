// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineItem renders as expected 1`] = `
<div>
  <div
    class="layerDirectionClass memo-layerDirectionBase-1"
  />
  <div
    class="layerOneDate"
    role="presentation"
  />
</div>
`;
